usecases/ast_eval/evaluate_ast.go

```diff
@@ -38,53 +80,83 @@ func EvaluateAst(ctx context.Context, environment AstEvaluationEnvironment, node
+		evaluated = true
+		weightedNodes := NewWeightedNodes(environment, node, node.Children)
 
-		// the frontend expects an ErrUndefinedFunction error to be present even when no evaluation happened.
-		if node.Function == ast.FUNC_UNDEFINED {
-			evaluation.Errors = append(evaluation.Errors, ast.ErrUndefinedFunction)
+		// eval each child
+		evaluation := ast.NodeEvaluation{
+			Index:         node.Index,
+			Function:      node.Function,
+			Children:      weightedNodes.Reorder(pure_utils.MapWhile(weightedNodes.Sorted(), evalChild)),
+			NamedChildren: pure_utils.MapValuesWhile(node.NamedChildren, evalChild),
 		}
 
-		return evaluation, false
-	}
+		if childEvaluationFail {
+			// an error occurred in at least one of the children. Stop the evaluation.
 
-	getReturnValue := func(e ast.NodeEvaluation) any { return e.ReturnValue }
-	arguments := ast.Arguments{
-		Args:      pure_utils.Map(evaluation.Children, getReturnValue),
-		NamedArgs: pure_utils.MapValues(evaluation.NamedChildren, getReturnValue),
-	}
+			// the frontend expects an ErrUndefinedFunction error to be present even when no evaluation happened.
+			if node.Function == ast.FUNC_UNDEFINED {
+				evaluation.Errors = append(evaluation.Errors, ast.ErrUndefinedFunction)
+			}
 
-	evaluator, err := environment.GetEvaluator(node.Function)
-	if err != nil {
-		evaluation.Errors = append(evaluation.Errors, err)
-		return evaluation, false
-	}
+			return nodeEvaluationResponse{evaluation, false}, nil
+		}
 
-	evaluation.ReturnValue, evaluation.Errors = evaluator.Evaluate(ctx, arguments)
+		getReturnValue := func(e ast.NodeEvaluation) any { return e.ReturnValue }
+		arguments := ast.Arguments{
+			Args:      pure_utils.Map(evaluation.Children, getReturnValue),
+			NamedArgs: pure_utils.MapValues(evaluation.NamedChildren, getReturnValue),
+		}
 
-	if evaluation.Errors == nil {
-		// Assign an empty array to indicate that the evaluation occured.
-		// The evaluator is not supposed to return a nil array of errors, but let's be nice.
-		evaluation.Errors = []error{}
-	}
+		evaluator, err := environment.GetEvaluator(node.Function)
+		if err != nil {
+			evaluation.Errors = append(evaluation.Errors, err)
+			return nodeEvaluationResponse{evaluation, false}, nil
+		}
+
+		evaluation.ReturnValue, evaluation.Errors = evaluator.Evaluate(ctx, arguments)
 
-	ok := len(evaluation.Errors) == 0
+		if evaluation.Errors == nil {
+			// Assign an empty array to indicate that the evaluation occured.
+			// The evaluator is not supposed to return a nil array of errors, but let's be nice.
+			evaluation.Errors = []error{}
+		}
+
+		ok := len(evaluation.Errors) == 0
+
+		if !ok {
+			// The evaluator is supposed to return nil ReturnValue when an error is present.
+			evaluation.ReturnValue = nil
+		}
+
+		evaluationResponse := nodeEvaluationResponse{evaluation, ok}
+
+		if cache != nil {
+			cache.Cache.Store(hash, deepcopy.Copy(evaluationResponse))
```

Comment:
Arguably, this copy should not be required.

We copy in two other instances:

* When retrieving this from the cache (since it may be shared by others using the cache)
* Below if other nodes were waiting for our result.